### ISSUE TITLE
server-side apply takes ownership of fields that don't get set due to status wiping

### ISSUE THREAD

**Issue by DirectXMan12** [kubernetes/kubernetes#75564]
The API server wipes status changes on calls to the normal endpoints (not /status) at the storage registry layer (right next to etcd).  This means that calls to `<resource>/<name>` can't change status.  However, if you send a server-side apply patch, the server-side apply logic still takes ownership of any fields you *try* to set, despite the fact that they don't get set.

Reproduced on HEAD (8af1bf313efeeafb94c01ce04965ad8fe9cea21a).

Reproducer: 
- Send a server-side apply patch setting any field in status to `/api/v1/namespaces/<ns>/pods/<name>`
- Observe that ownership of that field has been taken
- Observe that the field isn't set

cc #73723 
cc @jennybuckley @apelisse

**Comment by kwiesmueller** [comment]
After talking to @DirectXMan12 about this and digging into how status wiping works at the moment, I still got a few questions.
Primarily I would like to understand the reason why status wiping is happening where it is right now. Initially this seems to me like something that should be happening as for example, a mutating admission step.

@liggitt i was suggested to ask if you know more about this so I can understand the hows and whys. Ideally we would like to not own fields that get wiped but not prevent owning status fields entirely. If this happens further up the chain it might be easier to take action, for example the changes we currently suggest regarding the treatment of admission mutations would already take care of non-user changes to any field.

**Comment by liggitt** [comment]
> After talking to @DirectXMan12 about this and digging into how status wiping works at the moment, I still got a few questions.
> Primarily I would like to understand the reason why status wiping is happening where it is right now. Initially this seems to me like something that should be happening as for example, a mutating admission step.
> 
> @liggitt i was suggested to ask if you know more about this so I can understand the hows and whys. Ideally we would like to not own fields that get wiped but not prevent owning status fields entirely. If this happens further up the chain it might be easier to take action, for example the changes we currently suggest regarding the treatment of admission mutations would already take care of non-user changes to any field.

It happens in the REST handler strategy so that it is impossible for any changes to status to occur when modifying via the main resource (and vice-versa... it is impossible for changes to spec to occur via the status subresource). If this was done via admission early in the chain, a mutating webhook admission plugin could choose to make changes to spec or status via an unexpected endpoint

**Comment by kwiesmueller** [comment]
Thanks @liggitt we discussed this at the last api meeting and are now working on a solution or rather a KEP first to outline and discuss the approach.

**Comment by markyjackson-taulia** [comment]
@apelisse @kwiesmueller Bug triage for 1.17 here with a gentle reminder that code freeze for this release is on November 18. Is this issue still intended for 1.17?

**Comment by kwiesmueller** [comment]
@markyjackson-taulia there is a [WIP KEP PR](https://github.com/kubernetes/enhancements/pull/1123) addressing this issue and we are working on it. Looking at the amount of work still ahead and that the KEP wasn't really discussed outside the wg, I'm not sure if we'll make 1.17

**Comment by markyjackson-taulia** [comment]
@kwiesmueller ok let me know asap so I can adjust the milestone if needed

**Comment by markyjackson-taulia** [comment]
@kwiesmueller given there is still no work happening I am going to move this to milestone 1.18, are you ok with that?

**Comment by kwiesmueller** [comment]
Sure, we'll probably still be busy with the KEP PR for a while.
Sorry to keep you waiting

**Comment by smourapina** [comment]
Hello @liggitt  and @kwiesmueller! 
Bug Triage team here for the 1.18 release. This is a friendly reminder that code freeze is scheduled for 5 March, which is about 3 weeks from now. Is this issue still intended for milestone 1.18? Thanks in advance!

**Comment by kwiesmueller** [comment]
The KEP is merged and I plan to get this ready before code freeze. Yes.

**Comment by smourapina** [comment]
Hello again @kwiesmueller!
Just leaving another friendly ping to remind you that code freeze is coming in about 1 week (5 March), so please take this into account.

**Comment by kwiesmueller** [comment]
Thanks @smourapina .
We won't make it this release.

/milestone v1.19

**Comment by smourapina** [comment]
Thanks for the update @kwiesmueller, I will go ahead and change the milestone
/milestone v1.19

**Comment by liggitt** [comment]
is it reasonable to ship 1.18 with server-side apply on for all created objects by default if this isn't resolved?

**Comment by apelisse** [comment]
This bug won't have impact on operations on objects that are only updated (i.e. never applied): since updates are always forced, no conflicts are reported nor failing operations. The only impact for these objects is that they might have invalid information in their managed fields (which is *mostly fine*). The fact that we're enabling server-side apply for all created objects should only affect this use-case.

For objects that are applied though, there are mostly two cases:
1. A user/controller is trying to apply a field that they shouldn't apply (it's wiped). They might get a conflict on that field while the operation would normally go through. That sounds almost like a feature: I think we should fail when applying wiped fields.
2. A user/controller tries to apply a field that was NOT changed (because it was wiped) but we believe by mistake that it has changed (e.g. a controller has changed something in spec while sending a new status). If a controller is trying to apply, this is going to be fixed automatically since they mostly have to use force anyway. If a user is trying to apply, and the value hasn't changed, then they would be able to own the value anyway. They might get a conflict if they are trying to change the value, even if nobody had changed that value since they did. That's probably the most concerning part.

TL;DR: Main problem is: If a controller changes the spec while updating the status, a user might have a conflict when applying changes to that spec, even if they shouldn't.

Not sure how likely this is to happen, but it's equally likely to happen with 1.17.

**Comment by kbruner** [comment]
Hello from the bug triage team! This issue has not been updated for a long time, so I'd like to check what's the status. The code freeze is starting June 25th (about 4 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the issue is tagged for 1.19, is it still planned for this release?

**Comment by apelisse** [comment]
We're working on this and hoping to be done in 4 weeks. No guarantees though.

**Comment by kbruner** [comment]
@apelisse Is this still slated for 1.19? We're approaching code freeze.

**Comment by kwiesmueller** [comment]
I'd still like to get this in for 1.19, but don't know if it's realistic.
Even if we agree on a final implementation this week, it will still be a notable amount of work and take extensive review before we can merge it.
Very disappointing to not make it again.

**Comment by kbruner** [comment]
@kwiesmueller Thank you for the update. Can I go ahead and update the release milestone to 1.20?

**Comment by kbruner** [comment]
Hi @kwiesmueller. Just a reminder that the code freeze for 1.20 is tomorrow. Please let me know if I need to update the milestone for this issue to 1.20.

**Comment by kwiesmueller** [comment]
@kbruner you can update to 1.20 we didn't make it in time :-/

**Comment by liggitt** [comment]
as part of resolving this, remove the retry added in https://github.com/kubernetes/kubernetes/pull/94055 and add a time.Sleep(time.Second) to deterministically exercise the managedFields scenario

**Comment by kwiesmueller** [comment]
Thanks Jordan, I'll also add that to our tracking document so we don't miss it.

**Comment by MonzElmasry** [comment]
Hey :wave: Bug Triage here. Wanted to follow up on the status of this issue as we're approaching code freeze on 12.11.2020. This issue is tagged for 1.20, is it still planned for this release?

**Comment by kwiesmueller** [comment]
Yes, this is still intended for the next release.

**Comment by erismaster** [comment]
👋  Hi, Bug Triage here. Since https://github.com/kubernetes/kubernetes/pull/96498 is still draft and we're pretty late into code freeze, I am going to retarget this one for 1.21. Thanks!

/milestone v1.21

**Comment by apelisse** [comment]
Thanks for applying the label @erismaster, that's indeed our target :-)

**Comment by desponda** [comment]
Hello! I am from the Bug Triage team! Friendly reminder that code freeze is starting March 9th, 2021 (about 1 week from now). 
As the Issue is tagged for 1.21, is it still planned for this release based on the PR that was linked? The PR has not been marked for 1.21 Milestone

**Comment by apelisse** [comment]
Yes, we're working on it right now, and it's totally marked as 1.21.

**Comment by desponda** [comment]
Hi 👋 Bug Triage here, just a reminder that code freeze is end of day today! PRs need to be merged in before then, or they will require an approved exception to go into the 1.21 milestone.
